hnn_core/mpi_child.py

```diff
@@ -62,34 +62,37 @@ def _read_params(self):
             input_bytes = _read_all_bytes(sys.stdin.buffer)
             sys.stdin.close()
 
-            params = pickle.loads(codecs.decode(input_bytes, "base64"))
+            params = pickle.loads(base64.b64decode(input_bytes, validate=True))
         else:
             params = None
 
         params = self.comm.bcast(params, root=0)
         return params
 
+    def _pickle_data(self, sim_data):
+        # pickle the data and encode as base64 before sending to stderr
+        pickled_str = pickle.dumps(sim_data)
+        pickled_bytes = base64.b64encode(pickled_str)
+
+        return pickled_bytes
+
     def _write_data_stderr(self, sim_data):
         """write base64 encoded data to stderr"""
 
         # only have rank 0 write to stdout/stderr
         if self.rank > 0:
             return
 
-        # pickle the data and encode as base64 before sending to stderr
-        pickled_str = pickle.dumps(sim_data)
-        pickled_bytes = codecs.encode(pickled_str, 'base64')
+        pickled_bytes = self._pickle_data(sim_data)
 
-        # base64 encoding requires data padded to a multiple of 4
-        padding = len(pickled_bytes) % 4
-        pickled_bytes += b"===" * padding
         sys.stderr.write(pickled_bytes.decode())
         sys.stderr.flush()
 
-        # the parent process is waiting for "end_of_sim:[#bytes]" with the
-        # length of data
-        sys.stdout.write('end_of_data:%d' % len(pickled_bytes))
-        sys.stdout.flush()  # flush to ensure signal is not buffered
+        # the parent process is waiting for "@end_of_sim:[#bytes]@" with the
+        # length of data. The '@' is not found in base64 encoding, so we can
+        # be certain it is the border of the signal
+        sys.stderr.write('@end_of_data:%d@' % len(pickled_bytes))
+        sys.stderr.flush()  # flush to ensure signal is not buffered
```

Comment:
The signal is written to `stderr` (along with the data) instead of `stdout`. The `.flush()` calls ensure sequential ordering of (1) pickled data and (2) 'end_of_data` signal. I can't think of a scenario (3) would be more pickled data, but the code in `_read_stderr` would still work if that were the case.